spectec/src/backend-prose/render.ml

```diff
@@ -304,36 +304,47 @@ and render_iters env iters = List.map (render_iter env) iters |> List.fold_left
 
 (* Expressions and Paths *)
 
-(* fallback when expr cannot be embedded in a math block *)
+(* Invariant: All AL expressions fall into one of the three categories:
+  1. EL-like expression, only containing EL subexpressions
+  2. AL-only expression, possibly containing EL subexpressions 
+  3. pseudo-EL-like expression, containing at least one AL-only subexpression *)
+
+(* Category 1 is translated to EL then rendered by the Latex backend *)
+
+and render_expr env expr = match al_to_el_exp expr with
+  | Some exp -> 
+      (* embedded math blocks cannot have line-breaks *)
+      let newline = Str.regexp "\n" in
+      let sexp = Backend_latex.Render.render_exp env.render_latex exp in
+      let sexp = Str.global_replace newline "" sexp in
+      render_math sexp
+  | None -> render_expr' env expr
+
+(* Categories 2 and 3 are rendered by the prose backend,
+   yet EL subexpressions are still rendered by the Latex backend *)
 
 and render_expr' env expr =
   match expr.it with
-  | Al.Ast.VarE n | Al.Ast.SubE (n, _) -> n
-  | Al.Ast.NumE i -> Int64.to_string i
   | Al.Ast.BoolE b -> string_of_bool b
   | Al.Ast.UnE (NotOp, { it = Al.Ast.IsCaseOfE (e, kwd); _ }) ->
       let se = render_expr env e in
       let skwd = render_math (render_kwd env kwd) in
```

Comment:
Shouldn't this also be rendered as an EL atom, e.g., to ensure hints are used?

Reply:
Those cases belong to the *first* category, and are entirely handled in `render_expr` with `al_to_el_exp` translation. I believe there is no need to handle them in `render_expr'`, which concerns with categories other than the first.

Reply:
"Keywords" are user-defined atoms from the source. That is, they are part of the EL, in the same way other identifiers are. Like with other identifiers, we want a unified rendering for them. For example, the Latex renderer does some extra work to make atoms containing digits, e.g, i32, look nicer. It also should handle the cross-referencing for them in a single place (though not working yet, as you pointed out).

Reply:
You are definitely right, but I think the current implementation works just as what you have mentioned above. For AL `VarE` and `SubE`, they are translated to EL `VarE`, then rendered by the Latex backend. So in the spliced file, all meta-variables are rendered in a coherent fashion.

These expressions do not contain any subexpression, so they must be handled entirely in `al_to_el_exp`, which means that there is no need to write rendering logic for them in `render_expr'`.

```ocaml
  | Al.Ast.VarE id | Al.Ast.SubE (id, _) -> 
      let elid = id $ no_region in
      Some (El.Ast.VarE (elid, []))
``` 

Or maybe we are discussing in a different context. By "keyword", are you referring to this part of the implementation?

```ocaml
| Al.Ast.UnE (NotOp, { it = Al.Ast.IsCaseOfE (e, kwd); _ }) ->
      let se = render_expr env e in
      let skwd = render_math (render_kwd env kwd) in
```

Reply:
Yes, sorry for the confusion. What I meant is render_kwd specifically, and that it should render it through the Latex backend as an EL atom.